crates/sui-core/src/consensus_handler.rs

```diff
@@ -63,58 +63,74 @@ impl ConsensusHandler {
 impl ExecutionState for ConsensusHandler {
     /// This function will be called by Narwhal, after Narwhal sequenced this certificate.
     #[instrument(level = "trace", skip_all)]
-    async fn handle_consensus_transaction(
+    async fn handle_consensus_output(
         &self,
         // TODO [2533]: use this once integrating Narwhal reconfiguration
-        consensus_output: &Arc<ConsensusOutput>,
-        consensus_index: ExecutionIndices,
-        serialized_transaction: Vec<u8>,
+        consensus_output: ConsensusOutput,
     ) {
-        let index = Self::update_hash(
-            &self.last_seen,
-            consensus_index.clone(),
-            &serialized_transaction,
-        );
-        let index = if let Some(index) = index {
-            index
-        } else {
-            debug!(
+        let mut sequenced_transactions = Vec::new();
+        let mut seq = 0;
+
+        for (cert, batches) in consensus_output.batches {
+            let round = cert.header.round;
+            let author = cert.header.author.clone();
+            let output_cert = Arc::new(cert);
+            for batch in batches {
+                for serialized_transaction in batch.transactions {
+                    let transaction = match bincode::deserialize::<ConsensusTransaction>(
+                        &serialized_transaction,
+                    ) {
+                        Ok(transaction) => transaction,
+                        Err(err) => {
+                            warn!(
+                                    "Ignoring malformed transaction (failed to deserialize) from {}: {}",
+                                    author, err
+                                );
+                            continue;
+                        }
+                    };
+                    let index = ExecutionIndices {
+                        last_committed_round: round,
+                        sub_dag_index: consensus_output.sub_dag.sub_dag_index,
+                        transaction_index: seq,
+                    };
+
+                    let index_with_hash =
```

Comment:
You slightly change order between deserialization and update_hash here - previously we always called update_hash before deserialization, and in this implementation you skip `update_hash` if deserialization fails. I think original approach we always calculating hash makes more sense, so perhaps you can move it before deserialization?

Reply:
The reason I moved it to be after the deserialization is that right after the deserialization, we will next reorder the transactions, and so the re-hashing should subsequently follow the execution order, which is only formed now after deserialization. We can add a call to update the hash in the error branch of deserialization? Can you describe a bit more about what that would be used for?

Reply:
Ok, I think I understand the problem.

The hash is used to detect consensus forks (we have already found two bugs in narwhal that cause forks about a month ago thanks to that hash). As such it does not need to follow execution order, it is ok to just calculate hash based on the narwhal output, regardless on which order transactions are executed.

Ok, I guess in this case no need to make code more complicated then needed, we can change the logic the way you do it in this PR